File: app/security/grant_verifier.py
"""Offline verification for grant-profile-v1 compact JWS grants."""

from __future__ import annotations

import base64
import binascii
import hashlib
import json
import re
from collections.abc import Mapping, Sequence
from dataclasses import dataclass
from typing import Any

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.hazmat.primitives.asymmetric.utils import encode_dss_signature

from contracts.loader import load_trust_set


_P256_ORDER = int(
    "FFFFFFFF00000000FFFFFFFFFFFFFFFFBCE6FAADA7179E84F3B9CAC2FC632551", 16
)
_B64U_RE = re.compile(r"^[A-Za-z0-9_-]*$")
_UUIDV7_RE = re.compile(
    r"^[0-9a-f]{8}-[0-9a-f]{4}-7[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$"
)
_ID_RE = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._~-]{0,127}$")
_ISSUER = "urn:bridge-clean:commercial-control-plane"
_PROFILE = "urn:bridge-clean:grant-profile:v1"
_PURPOSE_CODES = {"installation-binding": "ib", "membership": "ms", "license": "le"}
_PROFILES = {
    "installation_grant": {
        "typ": "urn:bridge-clean:grant:installation:v1",
        "purpose": "installation-binding",
        "lifetime": 2_592_000,
        "grace": 604_800,
    },
    "creator_account_binding": {
        "typ": "urn:bridge-clean:grant:creator-binding:v1",
        "purpose": "installation-binding",
        "lifetime": 604_800,
        "grace": 259_200,
    },
    "membership_snapshot": {
        "typ": "urn:bridge-clean:grant:membership:v1",
        "purpose": "membership",
        "lifetime": 86_400,
        "grace": 259_200,
    },
    "license_entitlement": {
        "typ": "urn:bridge-clean:grant:license:v1",
        "purpose": "license",
        "lifetime": 86_400,
        "grace": 604_800,
    },
}
_ROLE_SCOPES = {
    "administrator": {
        "creator-binding:approve", "creator-binding:revoke", "installation-claim:issue",
        "installation:provision", "membership:manage", "provisioning:milestone",
        "runtime:existing-data", "runtime:licensed-processing",
    },
    "creator_operator": {"runtime:existing-data", "runtime:licensed-processing"},
    "delegated_installer": {"installation-claim:issue", "installation:provision", "provisioning:milestone"},
    "owner": {
        "commercial:manage", "creator-binding:approve", "creator-binding:revoke",
        "installation-claim:issue", "installation:provision", "membership:manage",
        "organization:manage", "provisioning:milestone", "runtime:existing-data",
        "runtime:licensed-processing",
    },
}
_COMMON_CLAIMS = {
    "aud", "exp", "grant_type", "iat", "installation_id", "installation_key_id",
    "installation_key_jkt", "iss", "jti", "nbf", "organization_id", "profile", "sub",
}
_EXTRA_CLAIMS = {
    "installation_grant": set(),
    "creator_account_binding": {"approval_id", "approval_revision", "creator_account_id"},
    "membership_snapshot": {
        "allowed_creator_account_ids", "ciam_issuer", "ciam_subject", "membership_id", "roles", "scopes",
    },
    "license_entitlement": {"entitlement_id", "features", "product_id"},
}


@dataclass(frozen=True, slots=True)
class GrantVerificationContext:
    """Local values required to verify one grant."""

    expected_grant_type: str
    expected_audience: str
    expected_organization_id: str
    expected_installation_id: str
    expected_installation_key_id: str
    expected_installation_key_jkt: str
    expected_subject: str
    verifier_time: int
    tombstones: Sequence[Mapping[str, Any]] = ()
    requested_operation: str = ""
    requested_creator_account_ids: Sequence[str] = ()


@dataclass(frozen=True, slots=True)
class GrantVerification:
    """A named grant-verification outcome."""

    valid: bool
    result: str
    time_state: str


def load_pinned_trust_set(
    path: str, *, environment: str = "production"
) -> dict[str, Any]:
    """Load the manifest-pinned trust set for grant verification."""

    return load_trust_set(path, environment=environment)


def trusted_signing_keys(trust_set: Mapping[str, Any]) -> dict[str, str]:
    """Return the purpose of every usable key in a trust set, keyed by identifier.

    Raises ValueError when the trust set does not provide usable signing keys.
    """

    return {kid: purpose for kid, (purpose, _) in _trusted_keys(trust_set).items()}


def _outcome(valid: bool, result: str, time_state: str = "invalid") -> GrantVerification:
    return GrantVerification(valid=valid, result=result, time_state=time_state)


def _b64u_decode(value: object) -> bytes:
    if not isinstance(value, str) or not _B64U_RE.fullmatch(value) or "=" in value:
        raise ValueError("noncanonical_base64url")
    try:
        decoded = base64.urlsafe_b64decode(value + "=" * ((4 - len(value) % 4) % 4))
    except (ValueError, binascii.Error) as exc:
        raise ValueError("noncanonical_base64url") from exc
    if base64.urlsafe_b64encode(decoded).rstrip(b"=").decode("ascii") != value:
        raise ValueError("noncanonical_base64url")
    return decoded


def _canonical_json(value: Any) -> bytes:
    def check(item: Any) -> None:
        if item is None or isinstance(item, float):
            raise ValueError("unsupported_json_value")
        if isinstance(item, dict):
            if not all(isinstance(key, str) for key in item):
                raise ValueError("non_string_key")
            for key, child in item.items():
                check(key)
                check(child)
        elif isinstance(item, list):
            for child in item:
                check(child)
        elif not isinstance(item, (str, int, bool)):
            raise ValueError("unsupported_json_value")

    check(value)
    return json.dumps(
        value, ensure_ascii=False, allow_nan=False, sort_keys=True, separators=(",", ":")
    ).encode("utf-8")


def _strict_json(data: bytes) -> dict[str, Any]:
    def unique(pairs: list[tuple[str, Any]]) -> dict[str, Any]:
        result: dict[str, Any] = {}
        for key, value in pairs:
            if key in result:
                raise ValueError("duplicate_json_member")
            result[key] = value
        return result

    try:
        decoded = json.loads(data.decode("utf-8"), object_pairs_hook=unique)
    except (UnicodeError, json.JSONDecodeError) as exc:
        raise ValueError("invalid_json") from exc
    if not isinstance(decoded, dict):
        raise ValueError("schema_invalid")
    return decoded


def _jwk_thumbprint(jwk: Mapping[str, Any]) -> bytes:
    bare = {name: jwk[name] for name in ("crv", "kty", "x", "y")}
    return hashlib.sha256(_canonical_json(bare)).digest()


def _public_key(jwk: Mapping[str, Any]) -> ec.EllipticCurvePublicKey:
    if set(jwk) != {"crv", "kid", "kty", "x", "y"}:
        raise ValueError("invalid_trust_set")
    if jwk.get("crv") != "P-256" or jwk.get("kty") != "EC":
        raise ValueError("invalid_trust_set")
    x = int.from_bytes(_b64u_decode(jwk["x"]), "big")
    y = int.from_bytes(_b64u_decode(jwk["y"]), "big")
    try:
        return ec.EllipticCurvePublicNumbers(x, y, ec.SECP256R1()).public_key()
    except ValueError as exc:
        raise ValueError("invalid_trust_set") from exc


def _trusted_keys(trust_set: Mapping[str, Any]) -> dict[str, tuple[str, ec.EllipticCurvePublicKey]]:
    entries = trust_set.get("keys")
    if not isinstance(entries, list) or not entries:
        raise ValueError("invalid_trust_set")
    keys: dict[str, tuple[str, ec.EllipticCurvePublicKey]] = {}
    for entry in entries:
        if not isinstance(entry, Mapping):
            raise ValueError("invalid_trust_set")
        purpose = entry.get("purpose")
        jwk = entry.get("jwk")
        thumbprint = entry.get("thumbprint")
        if purpose not in _PURPOSE_CODES or not isinstance(jwk, Mapping) or not isinstance(thumbprint, str):
            raise ValueError("invalid_trust_set")
        public_key = _public_key(jwk)
        thumbprint_bytes = _jwk_thumbprint(jwk)
        computed = base64.urlsafe_b64encode(thumbprint_bytes).rstrip(b"=").decode("ascii")
        kid = jwk.get("kid")
        expected_kid = (
            f"bc1.{_PURPOSE_CODES[purpose]}."
            f"{base64.urlsafe_b64encode(thumbprint_bytes[:16]).rstrip(b'=').decode('ascii')}"
        )
        if kid != expected_kid or thumbprint != computed or kid in keys:
            raise ValueError("invalid_trust_set")
        keys[kid] = (purpose, public_key)
    return keys


def _set_error(values: Any) -> str | None:
    if not isinstance(values, list) or not all(isinstance(item, str) for item in values):
        return "schema_invalid"
    if len(values) != len(set(values)):
        return "duplicate_set_value"
    if values != sorted(values, key=lambda item: item.encode("utf-8")):
        return "unsorted_set"
    return None


def verify_grant(
    token: str,
    *,
    context: GrantVerificationContext,
    trust_set: Mapping[str, Any],
) -> GrantVerification:
    """Verify one compact JWS grant against pinned local trust and bindings."""

    profile = _PROFILES.get(context.expected_grant_type)
    if profile is None:
        return _outcome(False, "unsupported_grant_type")
    if not isinstance(token, str) or len(token) > 16_384 or token.count(".") != 2:
        return _outcome(False, "invalid_compact_jws")
    try:
        header_segment, payload_segment, signature_segment = token.split(".")
        if len(header_segment) > 512:
            raise ValueError("header_too_large")
        header_bytes = _b64u_decode(header_segment)
        payload_bytes = _b64u_decode(payload_segment)
        signature = _b64u_decode(signature_segment)
        if len(payload_bytes) > 12_288:
            raise ValueError("payload_too_large")
        header = _strict_json(header_bytes)
        payload = _strict_json(payload_bytes)
    except (ValueError, UnicodeError, json.JSONDecodeError) as exc:
        return _outcome(False, str(exc))
    try:
        if _canonical_json(header) != header_bytes:
            return _outcome(False, "noncanonical_header")
        if _canonical_json(payload) != payload_bytes:
            return _outcome(False, "noncanonical_payload")
    except ValueError as exc:
        return _outcome(False, str(exc))
    if set(header) != {"alg", "kid", "typ"} or header.get("alg") != "ES256":
        return _outcome(False, "invalid_header")
    if payload.get("grant_type") != context.expected_grant_type:
        return _outcome(False, "grant_type_mismatch")
    if header.get("typ") != profile["typ"]:
        return _outcome(False, "typ_mismatch")
    if payload.get("aud") != context.expected_audience:
        return _outcome(False, "audience_mismatch")
    try:
        keys = _trusted_keys(trust_set)
    except ValueError as exc:
        return _outcome(False, str(exc))
    entry = keys.get(header.get("kid"))
    if entry is None:
        return _outcome(False, "unknown_kid")
    purpose, public_key = entry
    if purpose != profile["purpose"]:
        return _outcome(False, "wrong_key_purpose")
    if len(signature) == 64 and int.from_bytes(signature[32:], "big") > _P256_ORDER // 2:
        return _outcome(False, "high_s_signature")
    if not _verify_signature(public_key, f"{header_segment}.{payload_segment}".encode("ascii"), signature):
        return _outcome(False, "invalid_signature")
    if set(payload) != _COMMON_CLAIMS | _EXTRA_CLAIMS[context.expected_grant_type]:
        return _outcome(False, "schema_invalid")
    if payload["profile"] != _PROFILE or payload["iss"] != _ISSUER:
        return _outcome(False, "issuer_or_profile_mismatch")
    if not _UUIDV7_RE.fullmatch(str(payload["jti"])):
        return _outcome(False, "invalid_jti")
    if not all(isinstance(payload[name], int) and not isinstance(payload[name], bool) for name in ("iat", "nbf", "exp")):
        return _outcome(False, "invalid_numeric_date")
    if payload["iat"] != payload["nbf"] or payload["exp"] != payload["iat"] + profile["lifetime"]:
        return _outcome(False, "invalid_time_contract")
    for field in ("organization_id", "installation_id", "installation_key_id"):
        if not isinstance(payload[field], str) or not _ID_RE.fullmatch(payload[field]):
            return _outcome(False, "schema_invalid")
    if payload["organization_id"] != context.expected_organization_id:
        return _outcome(False, "organization_mismatch")
    if payload["installation_id"] != context.expected_installation_id:
        return _outcome(False, "installation_mismatch")
    if (
        payload["installation_key_id"] != context.expected_installation_key_id
        or payload["installation_key_jkt"] != context.expected_installation_key_jkt
    ):
        return _outcome(False, "installation_key_mismatch")
    if payload["sub"] != context.expected_subject:
        return _outcome(False, "subject_mismatch")
    type_outcome = _verify_type_specific(payload, context)
    if type_outcome is not None:
        return type_outcome
    if context.verifier_time + 60 < payload["nbf"]:
        return _outcome(False, "not_yet_valid")
    for tombstone in context.tombstones:
        if (
            tombstone.get("scope_type") == "jti"
            and tombstone.get("grant_jti") == payload["jti"]
            and tombstone.get("effective_at", context.verifier_time + 1) <= context.verifier_time
        ):
            return _outcome(False, "revoked")
    if context.verifier_time < payload["exp"]:
        return _outcome(True, "accepted", "current")
    if context.verifier_time < payload["exp"] + profile["grace"]:
        return _outcome(True, "accepted", "grace")
    return _outcome(False, "expired")


def _verify_signature(
    public_key: ec.EllipticCurvePublicKey, signing_input: bytes, signature: bytes
) -> bool:
    if len(signature) != 64:
        return False
    r = int.from_bytes(signature[:32], "big")
    s = int.from_bytes(signature[32:], "big")
    if not 1 <= r < _P256_ORDER or not 1 <= s < _P256_ORDER:
        return False
    try:
        public_key.verify(
            encode_dss_signature(r, s), signing_input, ec.ECDSA(hashes.SHA256())
        )
    except InvalidSignature:
        return False
    return True


def _verify_type_specific(
    payload: Mapping[str, Any], context: GrantVerificationContext
) -> GrantVerification | None:
    if context.expected_grant_type == "creator_account_binding":
        if not isinstance(payload["approval_revision"], int) or payload["approval_revision"] <= 0:
            return _outcome(False, "schema_invalid")
    elif context.expected_grant_type == "membership_snapshot":
        for field in ("roles", "scopes", "allowed_creator_account_ids"):
            error = _set_error(payload[field])
            if error:
                return _outcome(False, error)
        if len(payload["allowed_creator_account_ids"]) > 64:
            return _outcome(False, "account_cap_exceeded")
        if any(role not in _ROLE_SCOPES for role in payload["roles"]):
            return _outcome(False, "unknown_role")
        exact_scopes = sorted(set().union(*(_ROLE_SCOPES[role] for role in payload["roles"])))
        if payload["scopes"] != exact_scopes:
            return _outcome(False, "role_scope_mismatch")
        if context.requested_operation.startswith("runtime:") and payload["roles"] == ["delegated_installer"]:
            return _outcome(False, "runtime_role_denied")
        if any(account not in payload["allowed_creator_account_ids"] for account in context.requested_creator_account_ids):
            return _outcome(False, "creator_account_not_allowed")
    elif context.expected_grant_type == "license_entitlement":
        error = _set_error(payload["features"])
        if error:
            return _outcome(False, error)
        allowed = {"agent-capture", "analytics-processing", "command-execution", "ingestion"}
        if not payload["features"] or not set(payload["features"]) <= allowed or payload["product_id"] != "bridge-clean":
            return _outcome(False, "schema_invalid")
    return None

File: extension/transport/grant-verifier.mjs
// Offline grant-profile-v1 verification for the two installation-binding grants.
// Check order and result codes follow app/security/grant_verifier.py.

const ORDER = BigInt(
  "0xffffffff00000000ffffffffffffffffbce6faada7179e84f3b9cac2fc632551",
);
const UUIDV7 =
  /^[0-9a-f]{8}-[0-9a-f]{4}-7[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/u;
const ID = /^[A-Za-z0-9][A-Za-z0-9._~-]{0,127}$/u;
const B64U = /^[A-Za-z0-9_-]*$/u;
export const GRANT_ISSUER = "urn:bridge-clean:commercial-control-plane";
export const GRANT_PROFILE = "urn:bridge-clean:grant-profile:v1";
const PURPOSE_CODES = Object.freeze({
  "installation-binding": "ib",
  membership: "ms",
  license: "le",
});
export const GRANT_PROFILES = Object.freeze({
  installation_grant: Object.freeze({
    typ: "urn:bridge-clean:grant:installation:v1",
    purpose: "installation-binding",
    lifetime: 2_592_000,
    grace: 604_800,
    extra: Object.freeze([]),
  }),
  creator_account_binding: Object.freeze({
    typ: "urn:bridge-clean:grant:creator-binding:v1",
    purpose: "installation-binding",
    lifetime: 604_800,
    grace: 259_200,
    extra: Object.freeze([
      "approval_id",
      "approval_revision",
      "creator_account_id",
    ]),
  }),
});
const COMMON_CLAIMS = [
  "aud",
  "exp",
  "grant_type",
  "iat",
  "installation_id",
  "installation_key_id",
  "installation_key_jkt",
  "iss",
  "jti",
  "nbf",
  "organization_id",
  "profile",
  "sub",
];
const encoder = new TextEncoder();

export class GrantTrustError extends Error {
  constructor() {
    super("invalid_trust_set");
    this.name = "GrantTrustError";
    this.code = "invalid_trust_set";
  }
}

class Refusal extends Error {
  constructor(code) {
    super(code);
    this.code = code;
  }
}

const outcome = (valid, result, timeState = "invalid") =>
  Object.freeze({ valid, result, time_state: timeState });

const codePointLength = (value) =>
  value.length > 32_768 ? Infinity : [...value].length;

function b64uEncode(bytes) {
  let binary = "";
  for (const byte of bytes) binary += String.fromCharCode(byte);
  return btoa(binary).replace(/\+/gu, "-").replace(/\//gu, "_").replace(/=+$/u, "");
}

function b64uDecode(value) {
  if (typeof value !== "string" || !B64U.test(value))
    throw new Refusal("noncanonical_base64url");
  let decoded;
  try {
    decoded = Uint8Array.from(
      atob(
        value.replace(/-/gu, "+").replace(/_/gu, "/") +
          "=".repeat((4 - (value.length % 4)) % 4),
      ),
      (ch) => ch.charCodeAt(0),
    );
  } catch {
    throw new Refusal("noncanonical_base64url");
  }
  if (b64uEncode(decoded) !== value) throw new Refusal("noncanonical_base64url");
  return decoded;
}

// Floats, null, and integers outside the safe range are held as this marker so
// the canonical check can refuse them after parsing, as the reference does.
const UNSUPPORTED = Symbol("unsupported_json_value");

// Strict JSON: Python json.loads grammar with duplicate members refused when an
// object closes.
function strictJson(bytes) {
  let text;
  try {
    text = new TextDecoder("utf-8", { fatal: true, ignoreBOM: true }).decode(bytes);
  } catch {
    throw new Refusal("invalid_json");
  }
  let pos = 0;
  const fail = () => {
    throw new Refusal("invalid_json");
  };
  const space = () => {
    while (pos < text.length && " \t\n\r".includes(text[pos])) pos += 1;
  };
  const literal = (word, value) => {
    if (text.startsWith(word, pos)) {
      pos += word.length;
      return value;
    }
    return fail();
  };
  const string = () => {
    pos += 1;
    let result = "";
    for (;;) {
      if (pos >= text.length) fail();
      const ch = text[pos];
      if (ch === '"') {
        pos += 1;
        return result;
      }
      if (ch.charCodeAt(0) < 0x20) fail();
      if (ch !== "\\") {
        result += ch;
        pos += 1;
        continue;
      }
      const esc = text[pos + 1];
      const simple = { '"': '"', "\\": "\\", "/": "/", b: "\b", f: "\f", n: "\n", r: "\r", t: "\t" };
      if (Object.hasOwn(simple, esc)) {
        result += simple[esc];
        pos += 2;
      } else if (esc === "u" && /^[0-9a-fA-F]{4}$/u.test(text.slice(pos + 2, pos + 6))) {
        result += String.fromCharCode(parseInt(text.slice(pos + 2, pos + 6), 16));
        pos += 6;
      } else fail();
    }
  };
  const number = () => {
    const match = /-?(?:0|[1-9][0-9]*)(\.[0-9]+)?([eE][-+]?[0-9]+)?/uy;
    match.lastIndex = pos;
    const found = match.exec(text);
    if (!found) fail();
    pos = match.lastIndex;
    if (found[1] || found[2]) return UNSUPPORTED;
    const value = Number(found[0]);
    return Number.isSafeInteger(value) ? value : UNSUPPORTED;
  };
  const value = () => {
    space();
    const ch = text[pos];
    if (ch === "{") {
      pos += 1;
      const pairs = [];
      space();
      if (text[pos] === "}") pos += 1;
      else
        for (;;) {
          space();
          if (text[pos] !== '"') fail();
          const key = string();
          space();
          if (text[pos] !== ":") fail();
          pos += 1;
          pairs.push([key, value()]);
          space();
          if (text[pos] === ",") pos += 1;
          else if (text[pos] === "}") {
            pos += 1;
            break;
          } else fail();
        }
      const result = Object.create(null);
      for (const [key, item] of pairs) {
        if (Object.hasOwn(result, key)) throw new Refusal("duplicate_json_member");
        result[key] = item;
      }
      return result;
    }
    if (ch === "[") {
      pos += 1;
      const items = [];
      space();
      if (text[pos] === "]") pos += 1;
      else
        for (;;) {
          items.push(value());
          space();
          if (text[pos] === ",") pos += 1;
          else if (text[pos] === "]") {
            pos += 1;
            break;
          } else fail();
        }
      return items;
    }
    if (ch === '"') return string();
    if (ch === "t") return literal("true", true);
    if (ch === "f") return literal("false", false);
    if (ch === "n") return literal("null", null);
    if (ch === "N") return literal("NaN", UNSUPPORTED);
    if (ch === "I") return literal("Infinity", UNSUPPORTED);
    if (ch === "-" && text[pos + 1] === "I") return literal("-Infinity", UNSUPPORTED);
    return number();
  };
  const result = value();
  space();
  if (pos !== text.length) fail();
  if (!result || typeof result !== "object" || Array.isArray(result))
    throw new Refusal("schema_invalid");
  return result;
}

const isObject = (item) =>
  item !== null && typeof item === "object" && !Array.isArray(item);

function compareCodePoints(a, b) {
  const x = [...a];
  const y = [...b];
  for (let i = 0; i < Math.min(x.length, y.length); i += 1) {
    const d = x[i].codePointAt(0) - y[i].codePointAt(0);
    if (d) return d;
  }
  return x.length - y.length;
}

// Sorted keys, compact separators, non-ASCII kept literal.
function canonicalJson(item) {
  const write = (v) => {
    if (v === null || v === UNSUPPORTED) throw new Refusal("unsupported_json_value");
    if (typeof v === "string") {
      if (!v.isWellFormed()) throw new Refusal("unsupported_json_value");
      return JSON.stringify(v);
    }
    if (typeof v === "boolean") return String(v);
    if (typeof v === "number") return String(v);
    if (Array.isArray(v)) return `[${v.map(write).join(",")}]`;
    if (isObject(v))
      return `{${Object.keys(v)
        .sort(compareCodePoints)
        .map((k) => `${write(k)}:${write(v[k])}`)
        .join(",")}}`;
    throw new Refusal("unsupported_json_value");
  };
  return encoder.encode(write(item));
}

const sameBytes = (a, b) => a.length === b.length && a.every((x, i) => x === b[i]);

const toBigInt = (bytes) =>
  BigInt("0x" + [...bytes].map((x) => x.toString(16).padStart(2, "0")).join(""));

export async function jwkThumbprint(jwk) {
  const bare = JSON.stringify({ crv: jwk.crv, kty: jwk.kty, x: jwk.x, y: jwk.y });
  return new Uint8Array(await crypto.subtle.digest("SHA-256", encoder.encode(bare)));
}

const exactKeys = (object, names) =>
  isObject(object) &&
  Object.keys(object).length === names.length &&
  names.every((name) => Object.hasOwn(object, name));

/**
 * Load a packaged trust set. Refuses one not marked production-usable unless
 * allowNonProduction is set, which only test harnesses do.
 */
export async function loadGrantTrustSet(input, { allowNonProduction = false } = {}) {
  try {
    const trust = structuredClone(input);
    if (!isObject(trust)) throw new GrantTrustError();
    if (!allowNonProduction && trust.production_usable !== true) throw new GrantTrustError();
    if (typeof trust.profile !== "string" || !Array.isArray(trust.keys) || !trust.keys.length)
      throw new GrantTrustError();
    const keys = new Map();
    for (const entry of trust.keys) {
      if (!isObject(entry)) throw new GrantTrustError();
      const { purpose, jwk, thumbprint } = entry;
      if (!Object.hasOwn(PURPOSE_CODES, purpose) || typeof thumbprint !== "string")
        throw new GrantTrustError();
      if (!exactKeys(jwk, ["crv", "kid", "kty", "x", "y"]) || jwk.crv !== "P-256" || jwk.kty !== "EC")
        throw new GrantTrustError();
      if (b64uDecode(jwk.x).length !== 32 || b64uDecode(jwk.y).length !== 32)
        throw new GrantTrustError();
      const digest = await jwkThumbprint(jwk);
      const kid = `bc1.${PURPOSE_CODES[purpose]}.${b64uEncode(digest.slice(0, 16))}`;
      if (jwk.kid !== kid || thumbprint !== b64uEncode(digest) || keys.has(kid))
        throw new GrantTrustError();
      const key = await crypto.subtle.importKey(
        "jwk",
        { kty: "EC", crv: "P-256", x: jwk.x, y: jwk.y },
        { name: "ECDSA", namedCurve: "P-256" },
        false,
        ["verify"],
      );
      keys.set(kid, Object.freeze({ purpose, key }));
    }
    return Object.freeze({ lookup: (kid) => (typeof kid === "string" ? keys.get(kid) : undefined) });
  } catch {
    throw new GrantTrustError();
  }
}

/** Decode a grant payload without verifying it, to propose context values. */
export function peekGrantClaims(token) {
  try {
    if (typeof token !== "string" || token.length > 16_384) return null;
    const parts = token.split(".");
    if (parts.length !== 3) return null;
    return strictJson(b64uDecode(parts[1]));
  } catch {
    return null;
  }
}

/**
 * Verify one compact JWS grant. context: expectedGrantType, expectedAudience,
 * expectedOrganizationId, expectedInstallationId, expectedInstallationKeyId,
 * expectedInstallationKeyJkt, expectedSubject, verifierTime, tombstones.
 */
export async function verifyGrant(token, context, trust) {
  const profile = Object.hasOwn(GRANT_PROFILES, context.expectedGrantType)
    ? GRANT_PROFILES[context.expectedGrantType]
    : null;
  if (!profile) return outcome(false, "unsupported_grant_type");
  if (typeof token !== "string" || codePointLength(token) > 16_384 || token.split(".").length !== 3)
    return outcome(false, "invalid_compact_jws");
  const [headerSegment, payloadSegment, signatureSegment] = token.split(".");
  let header, payload, headerBytes, payloadBytes, signature;
  try {
    if (codePointLength(headerSegment) > 512) throw new Refusal("header_too_large");
    headerBytes = b64uDecode(headerSegment);
    payloadBytes = b64uDecode(payloadSegment);
    signature = b64uDecode(signatureSegment);
    if (payloadBytes.length > 12_288) throw new Refusal("payload_too_large");
    header = strictJson(headerBytes);
    payload = strictJson(payloadBytes);
    if (!sameBytes(canonicalJson(header), headerBytes)) return outcome(false, "noncanonical_header");
    if (!sameBytes(canonicalJson(payload), payloadBytes)) return outcome(false, "noncanonical_payload");
  } catch (error) {
    if (error instanceof Refusal) return outcome(false, error.code);
    return outcome(false, "invalid_json");
  }
  if (!exactKeys(header, ["alg", "kid", "typ"]) || header.alg !== "ES256")
    return outcome(false, "invalid_header");
  if (payload.grant_type !== context.expectedGrantType) return outcome(false, "grant_type_mismatch");
  if (header.typ !== profile.typ) return outcome(false, "typ_mismatch");
  if (payload.aud !== context.expectedAudience) return outcome(false, "audience_mismatch");
  const entry = trust?.lookup?.(header.kid);
  if (!entry) return outcome(false, "unknown_kid");
  if (entry.purpose !== profile.purpose) return outcome(false, "wrong_key_purpose");
  if (signature.length === 64 && toBigInt(signature.slice(32)) > ORDER / 2n)
    return outcome(false, "high_s_signature");
  if (!(await verifySignature(entry.key, `${headerSegment}.${payloadSegment}`, signature)))
    return outcome(false, "invalid_signature");
  if (!exactKeys(payload, [...COMMON_CLAIMS, ...profile.extra])) return outcome(false, "schema_invalid");
  if (payload.profile !== GRANT_PROFILE || payload.iss !== GRANT_ISSUER)
    return outcome(false, "issuer_or_profile_mismatch");
  if (typeof payload.jti !== "string" || !UUIDV7.test(payload.jti)) return outcome(false, "invalid_jti");
  if (!["iat", "nbf", "exp"].every((name) => Number.isSafeInteger(payload[name])))
    return outcome(false, "invalid_numeric_date");
  if (payload.iat !== payload.nbf || payload.exp !== payload.iat + profile.lifetime)
    return outcome(false, "invalid_time_contract");
  for (const field of ["organization_id", "installation_id", "installation_key_id"])
    if (typeof payload[field] !== "string" || !ID.test(payload[field])) return outcome(false, "schema_invalid");
  if (payload.organization_id !== context.expectedOrganizationId) return outcome(false, "organization_mismatch");
  if (payload.installation_id !== context.expectedInstallationId) return outcome(false, "installation_mismatch");
  if (
    payload.installation_key_id !== context.expectedInstallationKeyId ||
    payload.installation_key_jkt !== context.expectedInstallationKeyJkt
  )
    return outcome(false, "installation_key_mismatch");
  if (payload.sub !== context.expectedSubject) return outcome(false, "subject_mismatch");
  if (
    context.expectedGrantType === "creator_account_binding" &&
    !(Number.isSafeInteger(payload.approval_revision) && payload.approval_revision > 0)
  )
    return outcome(false, "schema_invalid");
  const now = context.verifierTime;
  if (!Number.isSafeInteger(now)) return outcome(false, "invalid_verifier_time");
  if (now + 60 < payload.nbf) return outcome(false, "not_yet_valid");
  for (const tombstone of context.tombstones ?? []) {
    const effective = Object.hasOwn(tombstone, "effective_at") ? tombstone.effective_at : now + 1;
    if (tombstone.scope_type === "jti" && tombstone.grant_jti === payload.jti && effective <= now)
      return outcome(false, "revoked");
  }
  if (now < payload.exp) return outcome(true, "accepted", "current");
  if (now < payload.exp + profile.grace) return outcome(true, "accepted", "grace");
  return outcome(false, "expired");
}

async function verifySignature(key, signingInput, signature) {
  if (signature.length !== 64) return false;
  const r = toBigInt(signature.slice(0, 32));
  const s = toBigInt(signature.slice(32));
  if (r < 1n || r >= ORDER || s < 1n || s >= ORDER) return false;
  try {
    return await crypto.subtle.verify(
      { name: "ECDSA", hash: "SHA-256" },
      key,
      signature,
      encoder.encode(signingInput),
    );
  } catch {
    return false;
  }
}
